crates/service/src/lib.rs

```diff
@@ -28,28 +32,71 @@ impl<'a> TerraphimService {
+        Ok(document)
     }
 
-    /// Create article
-    pub async fn create_article(&mut self, article: Article) -> Result<Article> {
-        self.config_state.index_article(&article).await?;
-        Ok(article)
+    /// Get the role for the given search query
+    async fn get_search_role(&self, search_query: &SearchQuery) -> Result<Role> {
+        let search_role = search_query.role.clone().unwrap_or_default();
+        let Some(role) = self.config_state.get_role(&search_role).await else {
+            return Err(ServiceError::Config(format!(
+                "Role {} not found in config",
+                search_role
+            )));
+        };
+        Ok(role)
     }
 
-    /// Search for articles in the haystacks
-    pub async fn search_articles(&self, search_query: &SearchQuery) -> Result<Vec<Article>> {
-        self.update_thesaurus(search_query).await?;
+    /// Search for documents in the haystacks
+    pub async fn search_documents(&self, search_query: &SearchQuery) -> Result<Vec<Document>> {
+        // Get the role from the config
+        log::debug!("Role for searching: {:?}", search_query.role);
+        let role = self.get_search_role(search_query).await?;
 
-        let cached_articles =
-            terraphim_middleware::search_haystacks(self.config_state.clone(), search_query.clone())
+        match role.relevance_function {
+            RelevanceFunction::TitleScorer => {
+                let index = terraphim_middleware::search_haystacks(
+                    self.config_state.clone(),
+                    search_query.clone(),
+                )
                 .await?;
-        let docs: Vec<IndexedDocument> = self.config_state.search_articles(search_query).await;
-        let articles = terraphim_types::merge_and_serialize(cached_articles, docs);
 
-        Ok(articles)
+                let indexed_docs: Vec<IndexedDocument> = self
+                    .config_state
+                    .search_indexed_documents(search_query)
+                    .await;
+
+                let documents = index.get_documents(indexed_docs);
+                // Sort the documents by relevance
+                let documents = score::sort_documents(search_query, documents);
+                Ok(documents)
+            }
+            RelevanceFunction::TerraphimGraph => {
+                self.build_thesaurus(search_query).await?;
+                let indexed_docs: Vec<IndexedDocument> = self
+                    .config_state
+                    .search_indexed_documents(search_query)
+                    .await;
+
+                // TODO: Convert indexed documents to documents
+                // We probably need to adjust the Thesaurus logseq haystack parser for this.
+                // let documents: Vec<Document> = indexed_docs
+                //     .iter()
+                //     .map(|indexed_doc| indexed_doc.to_document())
+                //     .collect();
+                todo!()
```

Comment:
@AlexMikhalev this is related to my previous comment. This is where we need to access the documents from the index and convert them back to real documents that we can show to the user.

Is there maybe an easier way?